File: iOS/TUILiveKit/Sources/Common/Basic/Pool/ListManager.swift
//
//  ListManager.swift
//  TUILiveKit
//
//  Created by gg on 2024/12/23.
//

import Foundation

class ListManager<T> {
    private var storage: [T] = []
    private let maxLength: Int
    private(set) var totalCount = 0
    
    var count: Int {
        storage.count
    }
    
    init(maxLength: Int = 1000) {
        self.maxLength = maxLength
    }
    
    func removeAll() {
        storage.removeAll()
    }
    
    @discardableResult
    func popFirst() -> T? {
        if storage.isEmpty {
            return nil
        }
        return storage.removeFirst()
    }
    
    func insert(_ obj: T, at index: Int) {
        totalCount += 1
        if count + 1 > maxLength {
            storage.removeFirst()
        }
        storage.insert(obj, at: index)
    }
    
    func append(_ obj: T) {
        totalCount += 1
        if count + 1 > maxLength {
            storage.removeFirst()
        }
        storage.append(obj)
    }
    
    func append(contentsOf list: [T]) {
        totalCount += list.count
        if list.count > maxLength {
            storage.append(contentsOf: list)
            storage.removeFirst(list.count - maxLength)
            return
        }
        if count + list.count > maxLength {
            let removeCount = count + list.count - maxLength
            storage.removeFirst(removeCount)
        }
        storage.append(contentsOf: list)
    }
    
    @discardableResult
    func remove(at index: Int) -> T? {
        if index >= storage.count || index < 0 {
            return nil
        }
        return storage.remove(at: index)
    }
    
    func reverse(index: Int) -> T? {
        if index < 0 || index >= storage.count {
            return nil
        }
        let reverseIndex = storage.count - 1 - index
        return storage[reverseIndex]
    }
    
    subscript(index: Int) -> T? {
        if index >= storage.count || index < 0 {
            return nil
        }
        return storage[index]
    }
}

extension ListManager {
    func getList() -> [T] {
        storage
    }
    
    func callAsFunction() -> [T] {
        getList()
    }
}

File: Flutter/live_uikit_gift/lib/widget/gift/display/animation/manager/list_manager.dart
class ListManager<T> {
  final int maxLength;
  final List<T> _storage = [];
  int _totalCount = 0;

  int get totalCount => _totalCount;
  int get count => _storage.length;

  ListManager({this.maxLength = 1000});

  void removeAll() {
    _storage.clear();
  }

  T? popFirst() {
    if (_storage.isEmpty) return null;
    return _storage.removeAt(0);
  }

  void insert(T obj, int index) {
    if (index >= count) {
      append(obj);
      return;
    }

    _totalCount++;
    if (_storage.length + 1 > maxLength) {
      _storage.removeAt(0);
    }
    _storage.insert(index, obj);
  }

  void append(T obj) {
    _totalCount++;
    if (_storage.length + 1 > maxLength) {
      _storage.removeAt(0);
    }
    _storage.add(obj);
  }

  void appendAll(List<T> list) {
    _totalCount += list.length;

    if (list.length > maxLength) {
      _storage.addAll(list.sublist(list.length - maxLength));
      return;
    }

    if (_storage.length + list.length > maxLength) {
      final removeCount = _storage.length + list.length - maxLength;
      _storage.removeRange(0, removeCount);
    }
    _storage.addAll(list);
  }

  T? removeAt(int index) {
    if (index < 0 || index >= _storage.length) return null;
    return _storage.removeAt(index);
  }

  T? reverse(int index) {
    if (index < 0 || index >= _storage.length) return null;
    final reverseIndex = _storage.length - 1 - index;
    return _storage[reverseIndex];
  }

  T? operator [](int index) {
    if (index < 0 || index >= _storage.length) return null;
    return _storage[index];
  }

  List<T> getList() => List<T>.unmodifiable(_storage);

  List<T> call() => getList();
}